SLE Micro: add smoke tests for transactional update

jlausuch opened the pull request:
Basic checks for transactional update operations that
are not tested anywhere else.
This test can be extended later on if needed adding additional
changes in initrd, grub, and performing more checks.

- Related ticket: https://progress.opensuse.org/issues/98538

jlausuch commented:
VR: https://openqa.suse.de/tests/overview?build=jlausuch%2Fos-autoinst-distri-opensuse%23sle_micro_bootloader&distri=sle-micro&version=5.1

mloviska commented:
LGTM, could you add also image tests, pls ?

Vogtinator commented:
This test is a bit weird. It's called `bootloader`, but does a lot of stuff which isn't related to bootloaders, and doesn't verify that any of the actions did what they're supposed to.

`setup-selinux` shouldn't be in here. In the case that SELinux is already enabled, it's a noop (especially with the `enable_selinux` module doing the same), and in the case that SELinux is not enabled yet, it'll have a big impact on following modules, making their results invalid.

`t-u bootloader` at the end is AFAICT not necessary, that's already done earlier and there shouldn't be any difference between those runs.

IMO this fits better into the `transactional/` directory, because it's not (SLE) Micro(OS) specific.

jsrain commented:
> This test is a bit weird. It's called `bootloader`, but does a lot of stuff which isn't related to bootloaders, and doesn't verify that any of the actions did what they're supposed to.

The test expects to verify that the t-u updates related to bootloader do not break the system - this is why I asked for it. Perhaps the verification could be improved, but some of the calls in the workflow showed that they can make system unbootable.
 
> `setup-selinux` shouldn't be in here. In the case that SELinux is already enabled, it's a noop (especially with the `enable_selinux` module doing the same), and in the case that SELinux is not enabled yet, it'll have a big impact on following modules, making their results invalid.

Good catch - Could the test be updated so that SELinux is not enabled during the initial installation?
 
> `t-u bootloader` at the end is AFAICT not necessary, that's already done earlier and there shouldn't be any difference between those runs.

In principle you are right - it should not make any difference, but the purpose is that it does not actually break the system. With recent bugs, in particular on s/390x, it showed that some of the calls in the test have the potential to make a machine unbootable.

> IMO this fits better into the `transactional/` directory, because it's not (SLE) Micro(OS) specific.

Cannot comment on this.

jlausuch commented:
Moved to transactional dir.
Not sure how to name this test... it basically tries to mess up bootloader... but I don't know a better name for it, sorry. 
I added a condition for SELinux. 

The idea is to have a separate test for this, specially for s390x. I can skip openSUSE if you consider it doesn't bring any benefit for x86_64...

Vogtinator commented on tests/transactional/bootloader.pm:
```diff
@@ -0,0 +1,42 @@
+    select_console 'root-console';
+
+    action('bootloader',    'Reinstall bootloader');
+    action('setup-selinux', 'Enable SELinux') if (script_run("selinuxenabled") != 0);
```
Is there a specific reason to do it in this test?

Simply scheduling this module after `enable_selinux` should have the same effect, especially because `setup-selinux` also does the same as `bootloader`.

mloviska replied:
IMHO it is about showing that re-installing bootloader itself without enabled selinux does not break the system on any architecture. On s390x images and unless the initrd gets rebuilt, system become unbootable after enabling selinux.

Vogtinator replied:
That sounds like this is more of a regression test for a very specific bug.

Can this be generalized for a use case which runs into the bug? I think that approach results in a better kind of module.

Vogtinator commented on tests/transactional/bootloader.pm:
```diff
@@ -0,0 +1,42 @@
+    assert_script_run('selinuxenabled');
+    action('grub.cfg',   'Regenerate grub.cfg');
+    action('initrd',     'Regenerate initrd');
+    action('kdump',      'Regenerate kdump');
```
`kdump` doesn't have an impact on the bootloader, it just regenerates the kdump initrd

Vogtinator commented:
What about calling this test something like `transactional_update-extra` or similar?

It currently runs commands which don't fit into the existing `transactional_update` module, and with that name it could also test `cleanup` while at it.

jlausuch commented:
> What about calling this test something like `transactional_update-extra` or similar?
> 
> It currently runs commands which don't fit into the existing `transactional_update` module, and with that name it could also test `cleanup` while at it.

That sounds more reasonable, thanks for the suggestion.

What if I remove selinux enablement in this test, and we make this test run always by default (no variable to enable it), in the selinux tests and the non-selinux tests?

jsrain commented:
> What if I remove selinux enablement in this test, and we make this test run always by default (no variable to enable it), in the selinux tests and the non-selinux tests?

Hmm, sounds like a good idea.

jlausuch commented:
Updated. Let me know if you agree and I will start VRs.

Vogtinator commented:
> Updated. Let me know if you agree and I will start VRs.

Looks good to me!

jlausuch commented:
* [SLE-Micro](https://openqa.suse.de/tests/overview?version=5.1&distri=sle-micro&build=jlausuch%2Fos-autoinst-distri-opensuse%23sle_micro_bootloader)

* [MicroOS](https://openqa.opensuse.org/tests/overview?distri=microos&version=Tumbleweed&build=jlausuch%2Fos-autoinst-distri-opensuse%23sle_micro_bootloader)

jlausuch commented:
Thank you.
I'll merge when VRs are done. O3 takes forever today...

jlausuch commented:
All the VRs are passing. Merging
